enclave-runtime/src/tls_ra/tls_ra_server.rs

```diff
@@ -0,0 +1,156 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! Implementation of the server part of the state provisioning.
+
+use super::{authentication::ClientAuth, Opcode, TcpHeader};
+use crate::{
+	attestation::create_ra_report_and_signature,
+	error::{Error as EnclaveError, Result as EnclaveResult},
+	ocall::OcallApi,
+	tls_ra::seal_handler::{SealHandler, UnsealStateAndKeys},
+};
+use itp_ocall_api::EnclaveAttestationOCallApi;
+use itp_sgx_crypto::{AesSeal, Rsa3072Seal};
+use itp_stf_state_handler::GlobalFileStateHandler;
+use itp_types::ShardIdentifier;
+use rustls::{ServerConfig, ServerSession, Stream};
+use sgx_types::*;
+use std::{
+	backtrace::{self, PrintFormat},
+	io::{Read, Write},
+	net::TcpStream,
+	sync::Arc,
+};
+/// Server part of the TCP-level connection and the underlying TLS-level session.
+///
+/// Includes a seal handler, which handles the reading part of the data to be sent.
+struct TlsServer<'a, StateAndKeyUnsealer> {
+	tls_stream: Stream<'a, ServerSession, TcpStream>,
+	seal_handler: StateAndKeyUnsealer,
+}
+
+impl<'a, StateAndKeyUnsealer> TlsServer<'a, StateAndKeyUnsealer>
+where
+	StateAndKeyUnsealer: UnsealStateAndKeys,
+{
+	fn new(
+		tls_stream: Stream<'a, ServerSession, TcpStream>,
+		seal_handler: StateAndKeyUnsealer,
+	) -> Self {
+		Self { tls_stream, seal_handler }
+	}
+
+	/// Sends all relevant data of the specific shard to the client.
+	fn write_shard(&mut self) -> EnclaveResult<()> {
+		let shard = self.read_shard()?;
+		self.write_all(&shard)
+	}
+
+	/// Read the shard of the state the client wants to receive.
+	fn read_shard(&mut self) -> EnclaveResult<ShardIdentifier> {
+		let mut shard_holder = ShardIdentifier::default();
+		let shard = shard_holder.as_fixed_bytes_mut();
+		self.tls_stream.read(shard)?;
+		Ok(shard.into())
+	}
+
+	/// Sends all relevant data to the client.
+	fn write_all(&mut self, shard: &ShardIdentifier) -> EnclaveResult<()> {
+		let shielding_key = self.seal_handler.unseal_shielding_key()?;
+		let signing_key = self.seal_handler.unseal_signing_key()?;
+		let state = self.seal_handler.unseal_state(shard)?;
```

Comment:
So this is probably new? We access the state here and write it to the TLS stream?